### ISSUE TITLE
Existing promotions feed schedule is not migrated from hourly to two-hour interval

### ISSUE THREAD

**Issue by bmzero** [facebook/facebook-for-woocommerce#4013]
## 🔎 Isolate the bug

- [ ] I have confirmed this occurs in the most recent version of WordPress, WooCommerce, and Meta for WooCommerce.
- [ ] I have confirmed this occurs when only WooCommerce and Meta for WooCommerce are active and when using a default WordPress or WooCommerce theme.

This was observed and corrected on a production store running WordPress 7.0.4, WooCommerce 11.0.0, and Meta for WooCommerce 3.7.6. I have not reproduced it on a minimal installation, but the stale recurrence is directly visible in Action Scheduler and follows from the scheduling logic described below.

## ✍️ Describe the bug

Meta for WooCommerce 3.7.6 defines the promotions feed generation interval as two hours:

```php
protected static function get_feed_gen_interval(): int {
    return 2 * HOUR_IN_SECONDS;
}
```

However, an existing `wc_facebook_regenerate_feed_promotions` recurring Action Scheduler record retained its older 3,600-second interval after the plugin was upgraded.

`AbstractFeed::schedule_feed_generation()` only creates the recurring action when none exists:

```php
if ( ! as_next_scheduled_action( $schedule_action_hook_name ) ) {
    as_schedule_recurring_action(
        time(),
        static::get_feed_gen_interval(),
        $schedule_action_hook_name,
        array(),
        facebook_for_woocommerce()->get_id_dasherized()
    );
}
```

It does not compare the existing action's recurrence with `get_feed_gen_interval()`. Consequently, stores upgrading from a version that scheduled this feed hourly can continue regenerating it hourly indefinitely, even though the installed code expects a two-hour cadence.

On a store with a large coupon catalog, each hourly launch produced approximately 149 `facebook_for_woocommerce/jobs/promotions_feed_generator/chain_batch` actions. WordPress then made repeated loopback requests to:

```text
/wp-admin/admin-ajax.php?action=as_async_request_queue_runner
```

Many queue-runner callbacks took several seconds, one exceeded twenty seconds, and the bursts correlated with origin-health latency alerts. There was no PHP-FPM worker-limit event during the examined incident.

This appears related to the Action Scheduler volume and loopback behavior reported in #3874, but the specific defect here is that an existing recurring schedule is not migrated when the plugin's intended interval changes.

## 🚶‍♀️ Steps to reproduce

1. Install a Meta for WooCommerce version that creates `wc_facebook_regenerate_feed_promotions` with a 3,600-second recurring interval, or create an equivalent recurring Action Scheduler action.
2. Upgrade to Meta for WooCommerce 3.7.6, where `PromotionsFeed::get_feed_gen_interval()` returns `2 * HOUR_IN_SECONDS`.
3. Allow the plugin and Action Scheduler to initialize normally.
4. Inspect the pending recurring action using Action Scheduler or the `actionscheduler_actions` table.
5. Observe that its serialized recurrence remains 3,600 seconds rather than changing to 7,200 seconds.
6. Allow the schedule to run and observe that promotions-feed generation still launches hourly.

As a production correction, I canceled only the stale recurring action and recreated it in the same `facebook-for-woocommerce` group with the same next-run timestamp and a 7,200-second interval. Readback confirmed the old action was canceled, the replacement was pending with the correct recurrence, and catalog synchronization remained enabled.

## ✔️ Expected behavior

When the plugin's intended feed-generation interval changes, existing recurring actions should be migrated to the current interval without requiring administrators to discover and replace stale Action Scheduler records manually.

A narrow fix could inspect the existing recurring action's schedule during plugin upgrade or feed initialization. If its recurrence differs from `static::get_feed_gen_interval()`, replace only that recurring action while preserving the next-run timestamp, arguments, and `facebook-for-woocommerce` group.

## 🗃 Logs

<details>
<summary>Sanitized Action Scheduler evidence</summary>

Environment:

```text
WordPress: 7.0.4
WooCommerce: 11.0.0
Meta for WooCommerce: 3.7.6
```

Before correction:

```text
hook: wc_facebook_regenerate_feed_promotions
status: pending
group: facebook-for-woocommerce
recurrence: 3600 seconds
installed-code interval: 7200 seconds
```

Observed workload from one launch:

```text
facebook_for_woocommerce/jobs/promotions_feed_generator/chain_batch: 149 completed actions
async endpoint: admin-ajax.php?action=as_async_request_queue_runner
slow callback range: several seconds, with one callback above 20 seconds
```

After replacing only the stale recurring action:

```text
old action: canceled
replacement action: pending
group: facebook-for-woocommerce
recurrence: 7200 seconds
catalog synchronization: still enabled
```

No credentials, customer data, request addresses, or private infrastructure details are included.

</details>

Related reports:

- #3874
- https://wordpress.org/support/topic/performance-issue-with-promotions-feed-generator-on-large-woocommerce-store/

**Comment by ukilla** [comment]
Hi @bmzero,

Thanks for the detailed report. Here is how to clear the stale schedule.

1. Go to WooCommerce > Status > Scheduled Actions, Pending tab.
2. Search for wc_facebook_regenerate_feed_promotions.
3. Check the Recurrence column. Every 2 hours is correct, Every 1 hour is stale.
4. Cancel the stale one. The plugin schedules a replacement at the current interval on the next hourly heartbeat, so check back within the hour and it should read Every 2 hours.

We recommend doing this during low traffic hours, because the replacement runs immediately and fires all of its `chain_batch` actions in one go, around 149 on your catalog. And don't cancel while the action is `in-progress`, because the finishing run schedules its own successor from the old recurrence and you can end up with two chains, hourly and two-hourly.

If the completed chain_batch rows have piled up, you can trim them with [WP-CLI](https://wp-cli.org/), the command line tool for WordPress. Run this from your WordPress root on the server:
`wp action-scheduler clean --status=complete,canceled --before='7 days ago'`
